### ISSUE TITLE
Umount networked mounts before cutting of the network connection

### ISSUE THREAD

**Issue by sjau** [systemd/systemd#5719]
### Submission type

  - [ x ] Request for enhancement (RFE)

### systemd version the issue has been seen with

systemd 232

### Used distribution

Nixos 17.09 unstable

### In case of bug report: Expected behaviour you didn't see

In run multiple VPN with OpenVPN and I mount different shares (some cifs) over the vpn.

### In case of bug report: Unexpected behaviour you saw

The problem is during shut down operation the vpn connection gets cut first and only afterwards it tries to umount the mounts. This causes a 180-300 second time while it's waiting for the non-responding mounts to respond.

### In case of bug report: Steps to reproduce the problem

- Install OpenVPN
- Mount a cifs network share through the vpn
- shut down the computer

**Comment by poettering** [comment]
Please mark network mounts with the "_netdev" mount option. systemd will then order them after "network.target", which — due to systemd's rule that the shutdown order is always the reverse of the startup order — means that the network mount will be shut down first, and the network only after that. For that to work correctly your networking solution (i.e. openvpn) needs to be ordered before network.target however. Also you need to run systemd 233 with current util-linux, since only then the information passing of "_netdev" from command line mount invocations gets properly propagated to systemd.

Closing this now. Feel free to reopen, if you can reproduce this with systemd 233 with current util-linux (specifically libmount), with the openvpn service properly ordered and with "_netdev" set as mount option correctly.

**Comment by ttulong** [comment]
Hi,
we are experiencing the same issue on Sles 12.3 (4.4) and RedHat 7.8 (3.10) with Rook Ceph 1.4.2 . Please let me know if you need anything else. Thanks

**Journalctl logs**
`-- Logs begin at Wed 2020-10-21 22:53:56 CDT, end at Fri 2020-10-30 08:46:58 CDT. --
Oct 29 22:42:08 xxxx kernel: libceph: mon1 10.43.175.61:6789 connect error
Oct 29 22:42:08 xxxxx kernel: libceph: connect 10.43.175.61:6789 error -101
Oct 29 22:42:04 xxxx kernel: libceph: mon1 10.43.175.61:6789 connect error
Oct 29 22:42:04 xxxx kernel: libceph: connect 10.43.175.61:6789 error -101

`
**mount logs**
`10.43.175.61:6789,10.43.93.185:6789,10.43.242.11:6789:/volumes/csi/csi-vol-c47b676c-f2df-11ea-9dcc-0a580a2a0bae/637920ff-bb45-4d4a-9bb6-2757fdff09e6 on /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-e34048d4-adae-4999-834f-06f64fd770d7/globalmount type ceph (rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=devops-cep fs,wsize=33554432,_netdev)
10.43.175.61:6789,10.43.93.185:6789,10.43.242.11:6789:/volumes/csi/csi-vol-c47f5490-f2df-11ea-9dcc-0a580a2a0bae/2278e388-c7bb-4a76-8e6e-5651d6a3bb69 on /var/lib/kubelet/pods/8c50c066-1ed4-47b4-b746-7500acfaddb3/volumes/kubernetes.io~csi/pvc-3f933360-34e4-43bf-bcd6-5bd1d566f4b9/mount type ceph (rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=devops-cephfs,wsize=33554432,_netdev)
10.43.175.61:6789,10.43.93.185:6789,10.43.242.11:6789:/volumes/csi/csi-vol-c47b676c-f2df-11ea-9dcc-0a580a2a0bae/637920ff-bb45-4d4a-9bb6-2757fdff09e6 on /var/lib/kubelet/pods/8c50c066-1ed4-47b4-b746-7500acfaddb3/volumes/kubernetes.io~csi/pvc-e34048d4-adae-4999-834f-06f64fd770d7/mount type ceph (rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=devops-cephfs,wsize=33554432,_netdev)
`

**Comment by zp841126** [comment]
Do we have any solutions for this issue. In my testing, when there are have 100 rbds, even if we added "_netdev",  it also encounter issue. It seems _netdev not functional as expected.

I go through systemd code, it will scan file /run/mount/utab and create unit, i think, the issue is create unit will fail in some corner case.

**Comment by ttulong** [comment]
hi @zp841126 , i still have the issue. I  manually drained the node and reboot.
There are some scripts that you can try to unmount all the drives at the reboot but that doesnt work for me neither.
 I see the error occurred after the journal logs : Failed to unmount /var/lib/docker

**Comment by zp841126** [comment]
I added some debug log to systemd  and found the issue caused by mount point name is too long

bool unit_name_is_valid(const char *n, enum template_valid template_ok) {
        const char *e, *i, *at;

        /* Valid formats:
         *
         *         string@instance.suffix
         *         string.suffix
         */

        assert(IN_SET(template_ok, TEMPLATE_VALID, TEMPLATE_INVALID));

        if (isempty(n))
                return false;

        if (strlen(n) >= UNIT_NAME_MAX)
                return false;


src/shared/unit-name.h:#define UNIT_NAME_MAX 256

**Comment by ttulong** [comment]
> I added some debug log to systemd and found the issue caused by mount point name is too long
> 
> bool unit_name_is_valid(const char *n, enum template_valid template_ok) {
> const char *e, *i, *at;
> 
> ```
>     /* Valid formats:
>      *
>      *         string@instance.suffix
>      *         string.suffix
>      */
> 
>     assert(IN_SET(template_ok, TEMPLATE_VALID, TEMPLATE_INVALID));
> 
>     if (isempty(n))
>             return false;
> 
>     if (strlen(n) >= UNIT_NAME_MAX)
>             return false;
> ```
> 
> src/shared/unit-name.h:#define UNIT_NAME_MAX 256

Cool @zp841126 , How did you enable the debug for systemd? Do you mind to share the URL too? I just want to see if we have the same issue. Thanks

**Comment by zp841126** [comment]
Hi @poettering ,

Do you mean systemd version less than 233 have known issue? 

Also you need to run systemd 233 with current util-linux, since only then the information passing of "_netdev" from command line mount invocations gets properly propagated to systemd.

**Comment by zp841126** [comment]
Hi @poettering ,

I checked latest systemd code, with my understanding,  option "_netdev" not working if mount point created by call mount command.

code path: mount_setup_unit -> mount_setup_new_unit

In this code path, it not call function mount_add_default_dependencies, it means we will not create dependencies between mount and network target. 

Could you confirm is it by design?

Thanks,

Zhen

**Comment by yuwata** [comment]
@zp841126 and @ttulong This issue was already closed. I have not followed your discussion in detail, but if there is some regression about handling `_netdev` option, then please open a new issue page. Otherwise, we easily forgot your discussed. 

In case of creating new issue, please check that your issue can be reproduced with recent systemd releases (currently, v248 or newer). Also, please provide detailed information about the mounts (e.g. /proc/1/mounts or /proc/1/mountinfo, generated .mount unit, relevant part of `systemd-analyze dump`, and so on).

**Comment by nixpanic** [comment]
It seems #20566 was opened for the new `_netdev` problem report.

**Comment by yuwata** [comment]
Yeah. And a fix for the issue is proposed in #20568, though it is not reviewed yet.
